contracts/toCKB-typescript/src/switch/auction_signertimeout/mod.rs

```diff
@@ -1,7 +1,210 @@
+            btc_lot_size.get_sudt_amount()
+        }
+        XChainKind::Eth => {
+            let eth_lot_size = input_data.get_eth_lot_size()?;
+            eth_lot_size.get_sudt_amount()
+        }
+    };
+
+    debug!("begin verify since");
+    let auction_time = verify_since()?;
+    debug!("begin verify input");
+    let inputs_xt_amount = verify_inputs(toCKB_lock_hash.as_ref(), lot_amount)?;
+    debug!("begin verify output");
+    verify_outputs(
+        input_data,
+        inputs_xt_amount,
+        auction_time,
+        toCKB_lock_hash.as_ref(),
+        lot_amount,
+    )?;
+
+    Ok(())
+}
+
+fn verify_since() -> Result<u64, Error> {
+    let since = load_input_since(0, Source::GroupInput).map_err(|_| Error::InputSinceInvalid)?;
+
+    if since & REMAIN_FLAGS_BITS != 0 // check flags is valid
+        || since & LOCK_TYPE_FLAG == 0 // check if it is relative_flag
+        || since & METRIC_TYPE_FLAG_MASK != SINCE_TYPE_TIMESTAMP
+    // check if it is timestamp value
+    {
+        return Err(Error::InputSinceInvalid);
+    }
+
+    let auction_time = since & VALUE_MASK;
+    Ok(auction_time)
+}
+
+fn verify_inputs(toCKB_lock_hash: &[u8], lot_amount: u128) -> Result<u128, Error> {
+    // inputs[0]: toCKB cell
+    // inputs[1:]: XT cell the bidder provides
+    // check XT cell on inputs
+    let inputs_amount = get_sum_sudt_amount(1, Source::Input, toCKB_lock_hash)?;
+
+    if inputs_amount < lot_amount {
+        return Err(Error::FundingNotEnough);
+    }
+    Ok(inputs_amount)
+}
+
+fn verify_outputs(
+    input_data: &ToCKBCellDataView,
+    inputs_xt_amount: u128,
+    auction_time: u64,
+    toCKB_lock_hash: &[u8],
+    lot_amount: u128,
+) -> Result<(), Error> {
+    /*
+        outputs[0]: bidder cell, receive collateral and ckb change
+        outputs[1]: liquidation trigger cell
+        outputs[2]: signer cell
+        outputs[3]: XT cell which redeemer receive
+        outputs[4:]: XT change
+    */
+
+    // check bidder cell
+    debug!("begin check bidder cell");
+    debug!(
+        "auction_time: {}, AUCTION_MAX_TIME: {}",
+        auction_time, AUCTION_MAX_TIME
+    );
+
+    let mut output_index = 0;
+    // - 1. check bidder lock
+    if load_cell_lock_hash(output_index, Source::Output)? != load_cell_lock_hash(1, Source::Input)?
+    {
+        return Err(Error::InvalidAuctionBidderCell);
+    }
+    debug!("1. check bidder lock success! ");
+
+    // expect paying ckb to bidder,trigger and signer
+    // cap of toCKB_cell ==  XT_CELL_CAPACITY + to_bidder + to_trigger + to_signer
+    let asset_collateral = load_cell_capacity(0, Source::GroupInput)? - XT_CELL_CAPACITY;
+    let mut to_bidder = asset_collateral;
+    let init_collateral = asset_collateral * AUCTION_INIT_PERCENT as u64 / 100;
+    if auction_time < AUCTION_MAX_TIME {
+        to_bidder =
+            init_collateral + (asset_collateral - init_collateral) / AUCTION_MAX_TIME * auction_time
```

Comment:
```suggestion
            init_collateral + (asset_collateral - init_collateral) * auction_time / AUCTION_MAX_TIME
```

we should always use `*` first to remain precision as much as possible.